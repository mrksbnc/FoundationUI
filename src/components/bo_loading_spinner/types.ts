import type { BoSize } from '@/data/bo_size.constant'
import type {
	BoLoaderTextPosition,
	BoLoaderVariant,
} from '@/data/loader.constant'

export type BoLoadingSpinnerProps = {
	size?: BoSize
	variant?: BoLoaderVariant
	customColor?: string
	loaderText?: string
	textPosition?: BoLoaderTextPosition
}
