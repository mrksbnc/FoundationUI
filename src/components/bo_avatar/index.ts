export * from './bo_avatar.contant';
export * from './bo_avatar.type';
export { default as BoAvatar } from './BoAvatar.vue';
