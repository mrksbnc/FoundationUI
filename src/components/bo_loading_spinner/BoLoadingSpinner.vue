<template>
	<div
		role="status"
		class="flex flex-row items-center gap-4"
	>
		<div
			role="status"
			:class="classes"
			aria-label="loader"
			:style="customColorStyle"
		></div>
		<span v-if="displayLoaderText">
			<BoText
				role="status"
				:text="loaderText ?? ''"
				aria-label="label"
				:color="BoTextColor.disabled"
				:font-family="BoFontFamily.graphik"
			/>
		</span>
	</div>
</template>

<script setup lang="ts">
import { BoSize } from '@/data/bo_size.constant';
import { StringUtils, TailwindUtils } from '@/utils';
import { computed, toRefs, type StyleValue } from 'vue';
import { BoFontFamily, BoTextColor } from '../bo_text';
import BoText from '../bo_text/BoText.vue';
import { BoLoadingSpinnerVariant } from './bo_loading_spinner.constant';
import type { BoLoadingSpinnerProps } from './bo_loading_spinner.type';

const props = withDefaults(defineProps<BoLoadingSpinnerProps>(), {
	size: () => BoSize.default,
	variant: () => BoLoadingSpinnerVariant.primary,
});

const { size, variant, loaderText, customColor } = toRefs(props);

const defaultClasses =
	/*tw*/ 'inline-block animate-spin rounded-full border-[2px] border-current border-t-transparent';

const displayLoaderText = computed<boolean>(() => {
	return !StringUtils.isEmpty(loaderText.value ?? '');
});

const loaderTailwindCssSizeClasses = computed<string>(() => {
	switch (size.value) {
		case BoSize.extra_small:
			return /*tw*/ 'w-2 h-2';
		case BoSize.small:
			return /*tw*/ 'w-3 h-3';
		case BoSize.extra_large:
			return /*tw*/ 'w-7 h-7';
		case BoSize.large:
			return /*tw*/ 'w-5 h-5';
		case BoSize.default:
		default:
			return /*tw*/ 'w-4 h-4';
	}
});

const customColorStyle = computed<StyleValue>(() => {
	if (!StringUtils.isEmpty(customColor.value ?? '')) {
		return {
			color: customColor.value,
		};
	}

	return {};
});

const tailwindCssVariantClasses = computed<string>(() => {
	if (!StringUtils.isEmpty(customColor.value ?? '')) {
		return '';
	}

	switch (variant.value) {
		case BoLoadingSpinnerVariant.secondary:
			return /*tw*/ 'text-gray-600';
		case BoLoadingSpinnerVariant.danger:
			return /*tw*/ 'text-red-600';
		case BoLoadingSpinnerVariant.warning:
			return /*tw*/ 'text-yellow-600';
		case BoLoadingSpinnerVariant.success:
			return /*tw*/ 'text-green-600';
		case BoLoadingSpinnerVariant.dark:
			return /*tw*/ 'text-black';
		case BoLoadingSpinnerVariant.white:
			return /*tw*/ 'text-white';
		case BoLoadingSpinnerVariant.primary:
		default:
			return /*tw*/ 'text-blue-600';
	}
});

const classes = computed<string>(() => {
	return TailwindUtils.merge(
		defaultClasses,
		loaderTailwindCssSizeClasses.value,
		tailwindCssVariantClasses.value,
	);
});
</script>
