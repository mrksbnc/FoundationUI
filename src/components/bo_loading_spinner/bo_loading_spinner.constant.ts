export enum BoLoadingSpinnerVariant {
	primary = 'primary',
	secondary = 'secondary',
	danger = 'danger',
	warning = 'warning',
	success = 'success',
	dark = 'dark',
	white = 'white',
}
