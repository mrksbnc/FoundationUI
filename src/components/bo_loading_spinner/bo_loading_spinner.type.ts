import type { BoSize } from '@/data/bo_size.constant';
import type { BoLoaderVariant } from '@/data/loader.constant';

export type BoLoadingSpinnerProps = {
	size?: BoSize;
	variant?: BoLoaderVariant;
	customColor?: string;
	loaderText?: string;
};
