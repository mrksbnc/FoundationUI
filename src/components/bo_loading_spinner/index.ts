export * from './bo_loading_spinner.type';
export { default as BoLoadingSpinner } from './BoLoadingSpinner.vue';
