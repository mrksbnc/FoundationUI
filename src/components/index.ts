export * from '@/components/bo_badge';
export * from '@/components/bo_button';
export * from '@/components/bo_icon';
export * from '@/components/bo_spinner';
