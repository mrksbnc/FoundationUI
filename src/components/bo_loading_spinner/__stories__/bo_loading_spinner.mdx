import { Canvas, Meta, Controls } from '@storybook/blocks';
import * as BoLoadingSpinnerStories from './bo_loading_spinner.stories';

<Meta
	title="Components/bo-loading-spinner"
	component={BoLoadingSpinnerStories.BoLoadingSpinner}
/>

# bo-loading-spinner

A loading spinner is a graphical representation of an object or action that is being processed.

## Basic usage

```html
<bo-loading-spinner />
```

## Props

- `size` - The size of the spinner (optional)
  - default: `BoSize.default`
- `variant` - The variant of the spinner (optional)
  - default: `BoLoadingSpinnerVariant.primary`
- `loaderText` - The text to be displayed in the loader (optional)
  - default: `''`
- `customColor` - A hex value to set as the color of the spinner (optional)
  - default: `undefined`

## Examples

<Canvas of={BoLoadingSpinnerStories.Example} />
<Controls of={BoLoadingSpinnerStories.Example} />

## Variants

To change the color of the component, you can use the predifened `BoLoadingSpinnerVariant` enum. The `BoLoadingSpinnerVariant` enum includes the following variants:

- `primary`
- `secondary`
- `danger`
- `warning`
- `success`
- `dark`
- `white`

<Canvas of={BoLoadingSpinnerStories.Variants} />

## Sizes

The `size` prop can be used to change the size of the spinner. The `size` prop should be a member of the `BoSize` enum.

- `extra-small`
- `small`
- `default`
- `large`
- `extra-large`

<Canvas of={BoLoadingSpinnerStories.Sizes} />

## Custom color

The `customColor` prop can be used to change the custom color of the spinner.
The `customColor` prop should be a string that represents the color in `hex` format.

<Canvas of={BoLoadingSpinnerStories.CustomColor} />

## With loader text

The `loaderText` prop can be used to display a text in the spinner.

<Canvas of={BoLoadingSpinnerStories.WithLoaderText} />
