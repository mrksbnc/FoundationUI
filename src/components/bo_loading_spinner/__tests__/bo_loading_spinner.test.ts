import { BoLoadingSpinner } from '@/components/bo_loading_spinner';
import { BoSize } from '@/data/bo_size.constant';
import { BoLoaderVariant } from '@/data/loader.constant';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let wrapper: ReturnType<typeof mount>;

beforeEach(() => {
	wrapper = mount(BoLoadingSpinner, {
		props: {
			size: BoSize.default,
			variant: BoLoaderVariant.primary,
		},
	});
});

describe('bo_loading_spinner.vue', () => {
	test('should match snapshot', () => {
		expect(wrapper.html()).toMatchSnapshot();
	});
});
