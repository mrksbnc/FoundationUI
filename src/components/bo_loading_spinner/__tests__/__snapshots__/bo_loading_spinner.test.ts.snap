// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bo_loading_spinner.vue > should match snapshot 1`] = `
"<div role="status" class="flex h-full w-full content-center items-center justify-center gap-2 flex-col">
  <div role="status" class="inline-block animate-spin rounded-full border-[2px] border-current border-t-transparent w-4 h-4 text-blue-600"></div>
  <!--v-if-->
</div>"
`;
